pynamodb/connection/base.py

```diff
@@ -1198,11 +1199,53 @@ def query(self,
         """
         operation_kwargs = {TABLE_NAME: table_name}
         name_placeholders = {}
+        expression_attribute_values = {}
+
+        # Create KeyConditionExpression
+        tbl = self.get_meta_table(table_name)
+        if tbl is None:
+            raise TableError("No such table: {0}".format(table_name))
+        if index_name:
+            hash_keyname = tbl.get_index_hash_keyname(index_name)
+            if not hash_keyname:
+                raise ValueError("No hash key attribute for index: {0}".format(index_name))
+        else:
+            hash_keyname = tbl.hash_keyname
+        # TODO: refactor to use condition expression context when it is available
+        pk_name = _substitute_names(hash_keyname, name_placeholders)
+        pk_value = self._get_value_placeholder(table_name, hash_keyname, hash_key, expression_attribute_values)
+        key_condition_expression =  "{0} = {1}".format(pk_name, pk_value)
+        if key_conditions is None or len(key_conditions) == 0:
```

Comment:
Is it possible to move these into methods, as the code is getting complex?

The key_conditions validation/transformation logic should be in one method that returns sk_value_list and sk_name. 

Then can we split up the if statements into other methods? or another method that will eventually returns a `key_condition_expression`. it is hard to follow the code the way it is right now.